# Solving Systems of Linear Equations with NumPy - Lab

## Introduction 

Now you've gathered all the required skills needed to solve systems of linear equations. You saw why there was need to calculate inverses of matrices, followed by matrix multiplication to figure out the values of unknown variables. 

The exercises in this lab present some problems that can be converted into a system of linear equations. 

## Objectives
You will be able to:

* Build a system of equations from given problem by creating matrix and vectors in NumPy
* Calculate the inverse and dot-product to get the values for unknown variables.
* Solve linear equations using `numpy.linalg.solve()`

## Exercise 1

A coffee shop is having a sale on coffee and tea. 

On day 1, 29 bags of coffee and 41 bags of tea were sold, for a total of 490 dollars.

On day 2, they sold 23 bags of coffee and 41 bags of tea, for which customers paid a total of 448 dollars.  

How much does each bag cost?

In [1]:
# Solution:
import numpy as np
A = np.array([[29,41],[23,41]])
b = [490,448]
np.linalg.solve(A,b)

array([7., 7.])

### Explain your answer in terms of prices
Both items cost $7 per bag.

## Exercise 2

The cost of admission to a popular music concert was 162 dollars for 12 children and 3 adults. 

The admission was 122 dollars for 8 children and 3 adults in another music concert. 

How much was the admission for each child and adult?

In [2]:
# Create a system of equations
A = np.array([[12,3],[8,3]])
b = np.array([162,122])
np.linalg.solve(A,b)


array([10., 14.])

### Describe your result
Children tickets cost 10 dollars per person, while adult pays $14.

## Exercise 3

You want to make a soup containing tomatoes, carrots and onions.

Suppose you don't know the exact mix to put in, but you know there are 7 individual pieces of vegetables, and there are twice as many tomatoes as onions, and that the 7 pieces of vegetables cost 5.25 USD in total. 
You also know that onions cost 0.5 USD each, tomatoes cost .75 USD and carrots cost 1.25 USD each.

Create a system of equations to find out exactly how many of each of the vegetables is in your soup.

In [3]:
# Create a system of equations
A = np.array([[1,1,1],[1,0,-2],[0.75,1.25,0.5]])
b = [7,0,5.25]
np.linalg.solve(A,b)

array([4., 1., 2.])

### Explain your answer
There are 4 tomatoes, 1 carrot and 2 onions in the soup.

## Exercise 4

A landlord owns 3 properties: a 1-bedroom, a 2-bedroom , and a 3-bedroom house. 

The total rent he receives is 1240 USD. 

He needs to make some repairs, where those repairs cost 10% of the 1-bedroom house’s rent. The 2-bedroom repairs cost 20% of the 2-bedroom rental price, and 30% of the 3-bedroom house's rent for its repairs.  The total repair bill for all three houses was 276 USD. 

The 3-bedroom house's rent is twice the 1-bedroom house’s rent. 

How much is the individual rent for three houses?

In [4]:
# Create a system of equations
A = np.array([[1,1,1],[0.1,0.2,0.3],[-2,0,1]])
b = np.array([1240,276,0])
np.linalg.solve(A,b)

array([280., 400., 560.])

### Explain your answer
The costs are 280, 400, and 560 respectively.

## Summary
In this lab, you learned how to use Numpy to solve linear equations by taking inverses and matrix multiplication and also using numpy's `.solve()`. You'll now take these skills forward and see how you can define a simple regression problem using linear algebra and solve it with Numpy. 